### PR TITLE
Users can register

### PR DESCRIPTION
**Pull request by seoulection**
-Users can register with their first name, last name, email address, and password (all required fields)
-If email address already exists in the database, then an error message appears
-After successful registration, users are redirected to a success page

### CODE REVIEW

File: ebai-back/app/controllers/users_controller.rb [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+
+  private
+    # Use callbacks to share common setup or constraints between actions.
+    def set_user
+      @user = User.find(params[:id])
+    end
+
+    # Only allow a trusted parameter "white list" through.
+    def user_params
+      params.require(:user).permit(:first_name, :last_name, :email, :password, :password_confirmation)
```
Comment (kkrull) [inline ebai-back/app/controllers/users_controller.rb]: Good - I'm glad you're being careful to not persist anything you don't intend to store.

File: ebai-front/tests/integration/SignupForm.spec.js [hunk 2]
```diff
@@ -0,0 +1,90 @@
...
+    it('shows an error message for incorrect signup', async () => {
+      const localVue = createLocalVue()
+      const router = mockRouter(localVue)
+      const wrapper = mount(SignupForm, {
+        localVue,
+        router
+      })
+
+      createUser.mockResolvedValueOnce(422)
+      const firstNameInput = wrapper.find('#firstName')
```
Comment (kkrull) [inline ebai-front/tests/integration/SignupForm.spec.js]: I see a bit of repetition here where you write the code to fill in the form.  It's ok to have a it of duplication in test code, but I might still extract some helper methods for the sake of improving readability / shortening the test method.  

Check out the [page object pattern](https://martinfowler.com/bliki/PageObject.html) when you get a chance.  You don't have to move to the full pattern right away, but extracting a few helper functions inside of your test file will get you part way there and put you in a good position to maintain this as the page gets more complex.

File: ebai-front/src/api/users.js [hunk 3]
```diff
@@ -0,0 +1,15 @@
...
+  const res = await axios.get('http://localhost:3000/users')
+  return res.data
+}
+
+export async function createUser(data) {
+    try {
+      const res = await axios.post('http://localhost:3000/users', data)
+      return res.status
+    } catch (e) {
+      return 422
```
Comment (kkrull) [inline ebai-front/src/api/users.js]: How would the caller tell if the request succeeded?  It might be better to return a Promise, so that they can either `await` or `catch` it and either show the successful response or show an error message, accordingly.

TL;DR - if you change `return 422` to something like `throw Error('Failed to create user')`.  Then the caller can `await/catch` and handle accordingly.

File: ebai-back/spec/controllers/users_controller_spec.rb [hunk 4]
```diff
@@ -0,0 +1,139 @@
...
+      end
+    end
+  end
+
+  describe "DELETE #destroy" do
+    it "destroys the requested user" do
+      user = User.create! valid_attributes
+      expect {
+        delete :destroy, params: {id: user.to_param}, session: valid_session
+      }.to change(User, :count).by(-1)
```
[the remark this answers is not in the document]
Comment (seoulection) [reply ebai-back/spec/controllers/users_controller_spec.rb]: These are the tests generated by Rails 😅. Should we just remove the methods/tests that are outside the scope of this project?

File: ebai-back/app/models/user.rb [hunk 5]
```diff
@@ -0,0 +1,8 @@
+class User < ApplicationRecord
+  has_secure_password
+
+  validates :first_name, presence: true
+  validates :last_name, presence: true
+  validates :email, presence: true, uniqueness: true
+  validates :password, presence: true
```
[the remark this answers is not in the document]
Comment (seoulection) [reply ebai-back/app/models/user.rb]: The password is already stored hashed and salted! The `has_secure_password` method takes care of that 😎

[the remark this answers is not in the document]
Comment (omarnyte) [reply ebai-back/app/models/user.rb]: I completely missed that. My bad!

**Comment by omarnyte** [comment]
Coming back around to approve.
